update reading of bluesky plan csv output

LukeFiddy opened the pull request:
- add ability to read metadata
- replace calculation of voltage increment with the metadata value
- add ability to select detector dimension to use in analysis
- update names of columns in bluesky plan output
- add selection of slit position column based on metadata
- update tests for these new features

Commit:
update read_bluesky_output
- add ability to read metadata
- replace calculation of voltage increment with the metadata value
- add ability to select detector dimension to use in analysis
- update names of columns in bluesky plan output
- add selection of slit position column based on metadata
- update tests for these new features

Commit:
change typing and add ignores to plots_test numpy arrays

Commit:
fix reading of dimension metadata

added strip function to remove the \n character

Commit:
add test for read_metadata

Commit:
update metadata for deleted slit positions

Commit:
update test for deleted slit positions

Commit:
fix typo

Commit:
add test for detector_dimension option

Commit:
add case for metadata test where metadata is formatted incorrectly

dan-fernandes commented on src/bimorph_mirror_analysis/__main__.py:
```diff
@@ -91,6 +97,7 @@ def calculate_optimal_voltages(
     max_consecutive_voltage_difference: int,
     baseline_voltage_scan: int = 0,
     slit_range: tuple[float, float] | None = None,
+    detector_dimension: str | None = None,
```
Could: Use `enum.Enum` rather than `str`

LukeFiddy replied:
converted to use enums

dan-fernandes commented on src/bimorph_mirror_analysis/read_file.py:
```diff
@@ -1,11 +1,52 @@
+                case "channels":
+                    metadata[key] = int(line.split(" ")[1])
+                case _:
+                    print("an error has occured when reading the csv metadata")
```
I don't think this case is necessary. In theory the CSV could be filled with extra metadata that isn't used in this analysis program.

LukeFiddy replied:
Good point. will need to keep the case but just replace the print statement with pass

LukeFiddy replied:
replaced the print statement with pass.

chnaged the testcase for when extra metadata is present

dan-fernandes commented on src/bimorph_mirror_analysis/read_file.py:
```diff
@@ -1,11 +1,52 @@
+                case _:
+                    print("an error has occured when reading the csv metadata")
+
+    return metadata
```
Should: Add metadata validation either here or where this function is called. This will make debugging erroneous CSVs easier.

LukeFiddy replied:
what kind of validatiaon do you envision?

LukeFiddy replied:
will add a check for each expected item of metadata before returning

LukeFiddy replied:
has been added

dan-fernandes commented on src/bimorph_mirror_analysis/read_file.py:
```diff
@@ -29,26 +72,30 @@ def read_bluesky_plan_output(
-    if abs(max_diff) > abs(min_diff):
-        voltage_increment = max_diff
+    if detector_dimension is not None:
+        assert detector_dimension in ["x", "y", "X", "Y"], "Invalid detector dimension"
```
Should: Replace `assert` with a raised exception.

There's an argument that asserts should only be used for testing. In certain circumstances they are optimised out of the code, which means they can't be relied upon for determining program flow.

LukeFiddy replied:
changed to raise a ValueError

Commit:
remove erroring on reading unrecognised metadata

Commit:
add testcase for extra metadata in test_read_metadata

Commit:
add stripping of commas in metadata

Commit:
replace assert with raise ValueError in read_bluesky_plan_output

Commit:
replace assert with raise ValueError in read_bluesky_plan_output

Commit:
change detector dimension handling from strings to enums

Commit:
add testcase for read_file detector_dimension ValueError

Commit:
add cli test for detector dimension options in calculate voltages

Commit:
update docstring for read_bluesky_ouput_plan return value

Commit:
fix generate_plots command to use the suppplied baseline voltage scan index

Commit:
fix indexing logic of the baseline voltage scan for negative indices in read_bluesky_plan_output

Commit:
fix typo for interaction_matrix

Commit:
fix usage of enum string value for column names and checking if is a proper state

Commit:
remove valuerror for enum not acceptable, error raised by enum itself earleir

Commit:
define __del__ method on plot objects to remove the figure from matplotlibs global state when the object is removed

Commit:
cover detector_dimension options with tests for generate plots command

Commit:
add assertion that the generate_tests command does not fail

Commit:
add validation for metadata

Commit:
format string over 2 lines

Commit:
refactor metadta test.

adds raise condition

reading and asserting the file is in a helper function to make easier readability with nested flow control being added
